Better unboxing strategy

alainfrisch opened the pull request:
This PR changes the criterion for deciding whether to unbox a local let binding `let x = e1 in e2`.

Previously, the criterion was based purely on the shape of `e1`.  Unboxing is performed if evaluating this expression necessarily forces some boxing, i.e. if all branches naturally return an unboxed number (which then needs to be boxed to be stored in `x`).  The same criterion is applied when `x` is a mutable variable, obtained by eliminating a local reference.  This strategy leads to two kinds of "bad situations":
- A reference whose initial expression cannot be determined to be an unboxable number (at the level of Clambda) will not be unboxed, which means that all assignment to it will require boxing (if the value is obtained from an operation giving an unboxed number).  This can be seen easily with `let x = ref nan in ...` (`nan` is obtained by reading an external symbol, the information about its type is lost), or `let x = ref (f ()) in ...`.  This can be quite bad and the workaround is to make it explicit that the value is a float through `let x = ref (0. +. f ()) in ...` which is ugly.
- When `e1` is something like `if econd then ethen else eelse` then both `ethen` else `eelse` must be able to naturally return an unboxed number in order for the whole let-binding to be unboxed.  For instance, if `eelse` is `nan` (or a function call, or accessing a function's parameter, etc), then the result of `ethen` ends up being boxed.  This is for a safety reason: because of GADTs, multiple branches could return values of different types, so one cannot unbox a branch only because another one returns, say, a float.  (See http://caml.inria.fr/mantis/view.php?id=7023 )

The current PR changes the criterion as follows:
- When the let binding corresponds to an eliminated reference, and the reference was statically on an unboxable type, we always unbox the identifier, even if this does not eliminate an allocation for the initial value itself (we will potentially avoid allocations on assignments).
- Otherwise, if the let-bound identifier is known to be of an unboxable type, we decide to unbox if at least one branch would otherwise need to box its result.  This is safe, I think, because if the result type of the branching construct is a basic type, then all branches must return the same basic type (even in presence of GADTs).

The testsuite has been extended with two tests illustrating how the two kinds of "bad situations" now become boxing free.

This is the first time where unboxing depends on typing information.  This means that we won't benefit from these improvements if the type becomes known only after inlining.

alainfrisch committed:
Keep track of mutable variables obtained by eliminating local references holding aun unboxable number type.  Then force unboxing of such variables to avoid repeated boxing every time the reference is assigned.

alainfrisch committed:
Keep the approx on all Llet nodes, not only variables.

alainfrisch committed:
Detect all let-bindings to unboxable floats.

alainfrisch committed:
When deciding whether to unbox a let binding, if one knows statically that the type of the bound expression is an unboxed number, relax the criterion: unbox if this benefits to at least one branch of the bound expression (instead of checking all branches).

alainfrisch committed:
Doc.

alainfrisch committed:
Simplify printlambda: do not show Pgenblock annotation.

alainfrisch committed:
Add test for better unboxing strategy.

bobzhang commented:
would you take this opportunity to use record instead of tuples in the lambda types?

alainfrisch commented:
> would you take this opportunity to use record instead of tuples in the lambda types?

Do you mean inline records for constructor arguments?  I don't think this is really related to this topic, so I'd better leave that for another proposal.

alainfrisch commented:
Rebased to trunk.  The plumbing for extra information required for the improved unboxing strategy does not go through flambda yet.

alainfrisch commented:
@chambart ported the approach of the original version to flambda: https://github.com/chambart/ocaml-1/pull/13/files

Pierre: would you feel like rebasing your work on this PR?

Everyone: I'd like to merge this soon enough since it is likely to conflict often with other changes (since it adds fields to two constructors in both lambda and clambda).  I don't think the changes are dangerous, and Pierre's original port shows that there is no blocking point to porting the approach to flambda.  So: is anyone opposed to merging this?

alainfrisch committed:
Fix for ocamlnat.

alainfrisch commented:
The failed check is the new test, in flambda mode, as expected.  Is there an easy way to disable some tests in the suite when flambda is on?

alainfrisch committed:
Disable new tests when flambda is enabled.

chambart commented on bytecomp/lambda.mli:
```diff
@@ -152,6 +152,9 @@ and comparison =
     Pgenarray | Paddrarray | Pintarray | Pfloatarray
 
+and block_kind =
+    Pgenblock | Pfloatblock | Pboxedintblock of boxed_integer
```
The name of the type and constructor is quite misleading as this is applied to things that are not blocs

alainfrisch replied:
Indeed, I'll do some renaming.

chambart commented:
My version is a bit more precise on the block representation: the fields are annotated separately with their type. Would you object to a PR on this branch to get a bit closer to that ?

By the way, I will continue to only annotate let_mutable right now. the rest might be for later.

chambart committed:
Annotate each field of a block separately

chambart committed:
Annotate Let_mutable with the kind of value

chambart committed:
Use the mutable_flag instead of let_kind for clambda let

chambart committed:
Also record type information for int

chambart committed:
Translcore propagates more block type information

alainfrisch commented:
> Would you object to a PR on this branch to get a bit closer to that ?

Sure.  Perhaps it's best to wait for the renaming to avoid conflicts. (EDIT: Renaming done.)

alainfrisch committed:
Renaming block_kind -> value_kind.

chambart committed:
Fix bind_tuples test for flambda

The intermediate floats for the last comparison are lifted, hence allocated

chambart committed:
Activate the test for float reference unboxing on flambda

chambart committed:
Test some mutable record field unboxing with flambda.

chambart committed:
Improve approximations using block kind

chambart committed:
Array assignement primitive specialisation using the approximation

chambart committed:
Activate all float unboxing tests on flambda

alainfrisch committed:
Changelog.

alainfrisch committed:
More compact printer and fix test.

alainfrisch commented:
Thanks @chambart !

Does anybody else  ( @xavierleroy, @mshinwell? )  want to review the whole PR?

alainfrisch commented:
@chambart : can you have a look at the tests which fail with flambda (tests/translprim/comparison_table.ml.reference, tests/translprim/ref_spec.ml.reference, tests/letrec/evaluation_order_1.ml)?

chambart committed:
Disable letrec evaluation order test

chambart committed:
Remove some Flambda specificities to translcore

This prevents some type annotations to be propagated to the middle-end,
but this is not really important as those only concern constants which
already provide all the usefull information about their shape.

alainfrisch commented:
Thanks @chambart .  Can you say a few words about why the flambda changes on this PR had an impact on the evaluation order of letrec?

chambart commented:
On 17/03/2016 23:21, Alain Frisch wrote:

> Thanks @chambart https://github.com/chambart . Can you say a few
> words about why the flambda changes on this PR had an impact on the
> evaluation order of letrec?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/ocaml/ocaml/pull/336#issuecomment-198107275
> 
> This was linked to the "if Config.flambda" on some translcore cases that
> prevented some constant cases
> from being represented as Lconst_ . This made the let-rec look a bit
> different which can change the
> evaluation order.
> I think that even if this doesn't occur anymore (I removed the tests) we
> should still keep the test like that.
> Those do not reflect any explicit choice of evaluation order, and the
> current one is particularly contrived
> to explain. And moreover this keep getting broken for completely
> 
> ## unrelated reasons.
> 
> Pierre

mshinwell commented:
I will try to have a look at it, but it will be next week (w/c 21st March).

mshinwell commented:
(In fact I would definitely like to review this, having just had a quick scan of it.)

alainfrisch commented:
@mshinwell Have you had a chance to review this PR?

alainfrisch commented:
@mshinwell Ping?

alainfrisch commented:
All tests went fine; two core devs worked on it; plenty of time before the release than will include the change.  So I went ahead and merged the branch.

gasche commented:
[MPR#7227](http://caml.inria.fr/mantis/view.php?id=7227) may be a regression caused by this branch -- but it is hard to tell as little information is available, besides a `tests/float-unboxing` failure message:

> Fatal error: exception Failure("float and int64 record; alloc = 56000")

alainfrisch commented:
This is specific to the flambda pipeline.  @chambart Can you have a look?
